### DISCUSSION TITLE
Local run (ready made tuned setup for some kind or other k8s local)

### DISCUSSION THREAD

**Discussion by dzmitry-lahoda** [fluxcd/flux2#1561] [Proposals]
I can run Terraform locally. Flux + Kustomize = Terrafrom on Kubernetes level. So it would be greate to have Flux as local tool.

**Comment by hiddeco** [comment]
This would require a lot more information about what you consider a "local tool", and how Flux and the controllers at present aren't a "local tool" (as the binaries can be run perfectly locally as long as some variables are correctly configured), before I would be able to give a sensible answer.

**Reply by stefanprodan** [reply]
You can run all things locally with Kubernetes Kind or Docker for Mac/Windows or minikube, etc

If you just care about validating your YAMLs without a Kubernetes cluster, then check out: https://github.com/fluxcd/flux2-kustomize-helm-example/blob/main/scripts/validate.sh

**Reply by dzmitry-lahoda** [reply]
Flux has Kustomization dependsOn and some variable substitution. Kustomize does not that. Also git includes are here. 

I run kustomize and produce yaml. I run flux and produce yaml, but with more power (dependsOn validation, variables, includes). 

Thank I can run local plan -so it checks if, probably that can be applied to cluster. And than push my local git changes - so real cluster flux runs. 

I get it - can run kind or setup go binaries locally, but using flux out of box as tool which extends kustomize into realm of terraform like features could be great.

**Reply by dzmitry-lahoda** [reply]
validate is good. but it does not grabs include (terraform modules from git repos). Also it is not extended to terrafrom plan (next kubectl command for some  kind of dry run)

**Reply by dzmitry-lahoda** [reply]
what validate.sh not doing, it does not takes flux kustomize(image/repository/whatever) files, and does not tries to use these files to run kustomize locally as it is in cluster at least partially (check that we are in the git references, may do some git clients, find all kustomize kustomiations (validate seems does it does not sources itself from flux cluster yamls), may be validating that images controllers policies will find proper yamls files for replacements).

**Reply by dzmitry-lahoda** [reply]
it is easy to have all credential in . folder for git and cloud provider and cr, but setupping local k8s and propagating that things into local cluster with flux will take time. so may be it is best option. but i have seen terraform - it gives much things `validate` with no cloud creds and `plan` without really having write rigths to cloud.
